core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java

```diff
@@ -208,12 +221,26 @@ public CreateIndexRequest settings(Map source) {
      *
      * @param type   The mapping type
      * @param source The mapping source
+     * @deprecated use {@link #mapping(String, BytesReference, XContentType)} to avoid content type detection
      */
+    @Deprecated
     public CreateIndexRequest mapping(String type, String source) {
+        return mapping(type, new BytesArray(source), XContentFactory.xContentType(source));
+    }
+
+    /**
+     * Adds mapping that will be added when the index gets created.
+     *
+     * @param type   The mapping type
+     * @param source The mapping source
+     * @param xContentType the content type of the mapping source
+     */
+    public CreateIndexRequest mapping(String type, BytesReference source, XContentType xContentType) {
```

Comment:
I think we should keep it out of the builder for now; someone can always call it themselves if they need it. I don't think SMILE/CBOR really make sense but we have it for the rest layer